src/lib/tls/sessions_sql/tls_session_manager_sql.cpp

```diff
@@ -18,200 +18,274 @@ namespace Botan::TLS {
 Session_Manager_SQL::Session_Manager_SQL(std::shared_ptr<SQL_Database> db,
                                          const std::string& passphrase,
                                          RandomNumberGenerator& rng,
-                                         size_t max_sessions,
-                                         std::chrono::seconds session_lifetime) :
+                                         size_t max_sessions) :
+   Session_Manager(rng),
    m_db(std::move(db)),
-   m_rng(rng),
-   m_max_sessions(max_sessions),
-   m_session_lifetime(session_lifetime)
+   m_max_sessions(max_sessions)
+   {
+   create_or_migrate_and_open(passphrase);
+   }
+
+void Session_Manager_SQL::create_or_migrate_and_open(const std::string& passphrase)
+   {
+   switch(detect_schema_revision())
+      {
+      case CORRUPTED:
+      case PRE_BOTAN_3_0:
+      case EMPTY:
+         // Legacy sessions before Botan 3.0 are simply dropped, no actual
+         // migration is implemented. Same for apparently corrupt databases.
+         m_db->exec("DROP TABLE IF EXISTS tls_sessions");
+         m_db->exec("DROP TABLE IF EXISTS tls_sessions_metadata");
+         create_with_latest_schema(passphrase, BOTAN_3_0);
+         break;
+      case BOTAN_3_0:
+         initialize_existing_database(passphrase);
+         break;
+      default:
+         throw Internal_Error("TLS session db has unknown database schema");
+      }
+   }
```

Comment:
SQL database migration is "lazy" in a sense that it removes the old tables and recreated new ones with an updated schema. The same mechanism is used when the database is found to be corrupted.